### PR TITLE
Fix NSFirewallEndpointAssociation and verify against realGCP

### PR DESCRIPTION
**Pull request by gemmahou**
### BRIEF Change description

Address https://github.com/GoogleCloudPlatform/k8s-config-connector/issues/6787

1. Updated NSFirewallEndpointAssociation test cases, re-gen logs against real and mockGCP
2. Fix direct API by using the correct reference types(`computev1beta1.ComputeNetworkRef` and `NetworkSecurityTLSInspectionPolicyRef`) so they can be resolved correctly.

#### Does this PR add something which needs to be 'release noted'?

```release-note
None
```

- [ ] Reviewer reviewed release note.

#### Additional documentation e.g., references, usage docs, etc.:

```docs

```

#### Intended Milestone

Please indicate the intended milestone. 
- [ ] Reviewer tagged PR with the actual milestone.

### Tests you have done

- [ ] Run `make ready-pr` to ensure this PR is ready for review.
- [ ] Perform necessary E2E testing for changed resources.

### CODE REVIEW

File: pkg/test/resourcefixture/testdata/basic/networksecurity/v1alpha1/networksecurityfirewallendpoint/networksecurityfirewallendpoint-maximal/_http_mock.log
Comment (ldanielmadariaga) [inline pkg/test/resourcefixture/testdata/basic/networksecurity/v1alpha1/networksecurityfirewallendpoint/networksecurityfirewallendpoint-maximal/_http_mock.log:1]: Why are we deleting the mock logs in maximal and minimal? We should either regenerate both real and mock GCP or delete the complete folders in favor of the -org and -project tests

File: pkg/test/resourcefixture/testdata/basic/networksecurity/v1alpha1/networksecurityfirewallendpoint/networksecurityfirewallendpoint-minimal/create.yaml
Comment (ldanielmadariaga) [inline pkg/test/resourcefixture/testdata/basic/networksecurity/v1alpha1/networksecurityfirewallendpoint/networksecurityfirewallendpoint-minimal/create.yaml:1]: Why are we deleting this case?

File: pkg/test/resourcefixture/testdata/basic/networksecurity/v1alpha1/networksecurityfirewallendpoint/networksecurityfirewallendpoint-project/update.yaml [hunk 1]
```diff
@@ -5,12 +5,9 @@ metadata:
 spec:
```
Comment (ldanielmadariaga) [inline pkg/test/resourcefixture/testdata/basic/networksecurity/v1alpha1/networksecurityfirewallendpoint/networksecurityfirewallendpoint-project/update.yaml:5]: IMHO we should prefer updating actual resource fields rather than labels. Ditto other test cases.

File: pkg/test/resourcefixture/testdata/basic/networksecurity/v1alpha1/networksecurityfirewallendpointassociation/firewallendpointassociation-basic/create.yaml
Comment (ldanielmadariaga) [inline pkg/test/resourcefixture/testdata/basic/networksecurity/v1alpha1/networksecurityfirewallendpointassociation/firewallendpointassociation-basic/create.yaml:1]: I don't see billingProjectID being set here (needed for project billing)

**Review by ldanielmadariaga — CHANGES_REQUESTED** [review]
Please update title and description this PR is changing many more things than mentioned

File: pkg/controller/direct/networksecurity/networksecurityfirewallendpoint_controller.go [hunk 2]
```diff
@@ -28,10 +27,9 @@ import (
 	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/registry"
 	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/structuredreporting"
 
+	networksecurity "cloud.google.com/go/networksecurity/apiv1"
```
Comment (ldanielmadariaga) [inline pkg/controller/direct/networksecurity/networksecurityfirewallendpoint_controller.go:30]: Do we have a JSON client this still seems to be a GRPC client based on http log traffic

**Review by ldanielmadariaga — CHANGES_REQUESTED** [review]

File: pkg/test/resourcefixture/testdata/basic/networksecurity/v1alpha1/networksecurityfirewallendpoint/networksecurityfirewallendpoint-maximal/_http_mock.log
Comment (gemmahou) [reply pkg/test/resourcefixture/testdata/basic/networksecurity/v1alpha1/networksecurityfirewallendpoint/networksecurityfirewallendpoint-maximal/_http_mock.log:1]: I deleted the whole -min and -max folder and replaced with -project and -org. I think how GH displays the change is confusing 🫤 

Please check this link:
https://github.com/gemmahou/k8s-config-connector/tree/c12f49efd816239075fc2f56a4ff8236d241dc25/pkg/test/resourcefixture/testdata/basic/networksecurity/v1alpha1/networksecurityfirewallendpoint

File: pkg/test/resourcefixture/testdata/basic/networksecurity/v1alpha1/networksecurityfirewallendpoint/networksecurityfirewallendpoint-minimal/create.yaml
Comment (gemmahou) [reply pkg/test/resourcefixture/testdata/basic/networksecurity/v1alpha1/networksecurityfirewallendpoint/networksecurityfirewallendpoint-minimal/create.yaml:1]: They are replaced with -project and -org tests.

File: pkg/test/resourcefixture/testdata/basic/networksecurity/v1alpha1/networksecurityfirewallendpoint/networksecurityfirewallendpoint-project/update.yaml (hunk 1 above)
Comment (gemmahou) [reply pkg/test/resourcefixture/testdata/basic/networksecurity/v1alpha1/networksecurityfirewallendpoint/networksecurityfirewallendpoint-project/update.yaml:5]: Will address in #12350. In addition to updating "labels", I'll test updating "billingProjectID" for org level endpoint.

According to [public doc](https://docs.cloud.google.com/firewall/docs/manage-firewall-endpoints#update), updatable fields are: "billingProjectID"(org level endpoint only), "description" and "labels". However, "description" is not populated when describing the resource(see realGCP log and below gcloud output), thus I had a special handling in the direct controller compare logic: [code](https://github.com/gemmahou/k8s-config-connector/blob/7af8a755234f024781fb8c83f0dc2757c9b7aa98/pkg/controller/direct/networksecurity/networksecurityfirewallendpoint_controller.go#L251-L252) to ignore "description", and updating "description" may not be necessary.
 
```
gcloud network-security firewall-endpoints describe fwe-7a6fxqxcgyduqzi \
    --organization 128653134652 \
    --location us-central1-a
Parsed [firewall endpoint] resource: organizations/128653134652/locations/us-central1-a/firewallEndpoints/fwe-7a6fxqxcgyduqzi
associatedNetworks:
- projects/402887772785/global/networks/default
associations:
- name: projects/402887772785/locations/us-central1-a/firewallEndpointAssociations/assoc-abfef25f-6bf6-4fb1-bb38-2a72020166a9
  network: projects/402887772785/global/networks/default
billingProjectId: cnrm-yuhou
createTime: '2026-08-12T16:48:22.568869462Z'
endpointSettings:
  jumboFramesEnabled: true
labels:
  foo: bar
name: organizations/128653134652/locations/us-central1-a/firewallEndpoints/fwe-7a6fxqxcgyduqzi
state: ACTIVE
updateTime: '2026-08-15T03:28:10.443842666Z'
```

File: pkg/test/resourcefixture/testdata/basic/networksecurity/v1alpha1/networksecurityfirewallendpointassociation/firewallendpointassociation-basic/create.yaml
Comment (gemmahou) [reply pkg/test/resourcefixture/testdata/basic/networksecurity/v1alpha1/networksecurityfirewallendpointassociation/firewallendpointassociation-basic/create.yaml:1]: [billingProjectID](https://docs.cloud.google.com/firewall/docs/reference/network-security/rest/v1/projects.locations.firewallEndpoints#:~:text=is%20fully%20configured.-,billingProjectId,-string) is only needed for **org level NetworkSecurityFirewallEndpoint**, not for **NetworkSecurityFirewallEndpointAssociation**.

We have test coverage for project and org level NetworkSecurityFirewallEndpoint in #12350

File: pkg/controller/direct/networksecurity/networksecurityfirewallendpoint_controller.go (hunk 2 above)
Comment (gemmahou) [reply pkg/controller/direct/networksecurity/networksecurityfirewallendpoint_controller.go:30]: I did find a REST client: https://github.com/googleapis/google-cloud-go/blob/networksecurity/v0.19.0/networksecurity/apiv1/firewall_activation_client.go#L604

Will address in #12350

File: config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_networksecurityfirewallendpointassociations.networksecurity.cnrm.cloud.google.com.yaml [hunk 3]
```diff
@@ -189,15 +189,15 @@ spec:
                   - external
                 properties:
                   external:
-                    description: The TLS inspection policy selflink, when not managed
-                      by Config Connector.
+                    description: A reference to an externally managed NetworkSecurityTLSInspectionPolicy
+                      resource. Should be in the format "projects/{{projectID}}/locations/{{location}}/tlsInspectionPolicies/{{tlsInspectionPolicyID}}".
```
Comment (ldanielmadariaga) [inline config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_networksecurityfirewallendpointassociations.networksecurity.cnrm.cloud.google.com.yaml:193]: May be wrong but IIRC most CRDs don't use double curly braces for refs patterns.

File: config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_networksecuritygatewaysecuritypolicies.networksecurity.cnrm.cloud.google.com.yaml [hunk 4]
```diff
@@ -126,15 +126,15 @@ spec:
                   - external
                 properties:
                   external:
-                    description: The TLS inspection policy selflink, when not managed
-                      by Config Connector.
+                    description: A reference to an externally managed NetworkSecurityTLSInspectionPolicy
+                      resource. Should be in the format "projects/{{projectID}}/locations/{{location}}/tlsInspectionPolicies/{{tlsInspectionPolicyID}}".
```
Comment (ldanielmadariaga) [inline config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_networksecuritygatewaysecuritypolicies.networksecurity.cnrm.cloud.google.com.yaml:130]: May be wrong but IIRC most CRDs don't use double curly braces for refs patterns.

File: config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_networksecurityfirewallendpointassociations.networksecurity.cnrm.cloud.google.com.yaml [hunk 5]
```diff
@@ -124,14 +124,14 @@ spec:
                   - external
                 properties:
                   external:
-                    description: The network selflink of form "projects/{{project}}/global/networks/{{name}}",
-                      when not managed by Config Connector.
+                    description: A reference to an externally managed ComputeNetwork
+                      resource. Should be in the format "projects/{{projectID}}/global/networks/{{networkID}}".
```
Comment (ldanielmadariaga) [inline config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_networksecurityfirewallendpointassociations.networksecurity.cnrm.cloud.google.com.yaml:128]: May be wrong but IIRC most CRDs don't use double curly braces for refs patterns.

**Review by ldanielmadariaga — DISMISSED** [review]
Please update the description of this PR explaining what it actually does

File: config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_networksecurityfirewallendpointassociations.networksecurity.cnrm.cloud.google.com.yaml (hunk 5 above)
Comment (gemmahou) [reply config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_networksecurityfirewallendpointassociations.networksecurity.cnrm.cloud.google.com.yaml:128]: The description comes from https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/apis/compute/refs/computenetwork_reference.go#L49, so I believe all ComputeNetworkRef share the same format.

Search result: https://github.com/search?q=repo%3AGoogleCloudPlatform%2Fk8s-config-connector+projects%2F%7B%7BprojectID%7D%7D%2Fglobal%2Fnetworks%2F%7B%7BnetworkID%7D%7D&type=code

File: config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_networksecurityfirewallendpointassociations.networksecurity.cnrm.cloud.google.com.yaml (hunk 3 above)
Comment (gemmahou) [reply config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_networksecurityfirewallendpointassociations.networksecurity.cnrm.cloud.google.com.yaml:193]: We use double curly braces in other refs(see below ComputeNetworkRef), so I'll leave it as is for consistency.

File: config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_networksecuritygatewaysecuritypolicies.networksecurity.cnrm.cloud.google.com.yaml (hunk 4 above)
Comment (gemmahou) [reply config/crds/resources/apiextensions.k8s.io_v1_customresourcedefinition_networksecuritygatewaysecuritypolicies.networksecurity.cnrm.cloud.google.com.yaml:130]: Same as above comment, I'll leave it as is for consistency.

File: dev/tasks/build-release-bundle [hunk 6]
```diff
@@ -200,7 +200,6 @@ BLOCKED_KINDS=(
     "NetworkSecurityAddressGroup"
     "NetworkSecurityBackendAuthenticationConfig"
     "NetworkSecurityDNSThreatDetector"
-    "NetworkSecurityFirewallEndpoint"
```
Comment (maqiuyujoyce) [inline dev/tasks/build-release-bundle:202]: Why is this item removed here?

File: pkg/test/resourcefixture/testdata/basic/networksecurity/v1alpha1/networksecurityfirewallendpointassociation/firewallendpointassociation-basic/_http_mock.log [hunk 7]
```diff
@@ -14,25 +15,257 @@ X-Xss-Protection: 0
...
+  },
+  "name": "projects/${projectId}/locations/us-central1-a/operations/${operationID}",
+  "response": {
+    "@type": "type.googleapis.com/google.cloud.networksecurity.v1.FirewallEndpoint",
+    "billingProjectId": "${projectId}",
+    "createTime": "2024-04-01T12:34:56.123456Z",
+    "endpointSettings": {
+      "jumboFramesEnabled": true
+    },
+    "labels": {
```
Comment (maqiuyujoyce) [inline pkg/test/resourcefixture/testdata/basic/networksecurity/v1alpha1/networksecurityfirewallendpointassociation/firewallendpointassociation-basic/_http_mock.log:223]: Nit: the response from real GCP doesn't have labels here.

Not blocking the PR, but I think we should keep track of "weird" behaviors in operations.

File: pkg/test/resourcefixture/testdata/basic/networksecurity/v1alpha1/networksecurityfirewallendpointassociation/firewallendpointassociation-basic/_http_mock.log [hunk 8]
```diff
@@ -277,21 +510,213 @@ X-Xss-Protection: 0
     "apiVersion": "v1",
     "createTime": "2024-04-01T12:34:56.123456Z",
     "endTime": "2024-04-01T12:34:56.123456Z",
-    "target": "projects/${projectId}/locations/us-central1/firewallEndpointAssociations/fwea-${uniqueId}",
+    "target": "projects/${projectId}/locations/us-central1-a/firewallEndpointAssociations/fwea-${uniqueId}",
     "verb": "delete"
   },
-  "name": "projects/${projectId}/locations/us-central1/operations/${operationID}",
+  "name": "projects/${projectId}/locations/us-central1-a/operations/${operationID}",
   "response": {
```
Comment (maqiuyujoyce) [inline pkg/test/resourcefixture/testdata/basic/networksecurity/v1alpha1/networksecurityfirewallendpointassociation/firewallendpointassociation-basic/_http_mock.log:514]: The response section of a delete operation is an empty struct here in real GCP.

I think that is probably a better behavior in mock as well.

File: pkg/test/resourcefixture/testdata/basic/networksecurity/v1alpha1/networksecurityfirewallendpointassociation/firewallendpointassociation-basic/_http_mock.log [hunk 9]
```diff
@@ -277,21 +510,213 @@ X-Xss-Protection: 0
...
+Content-Type: application/json; charset=UTF-8
+Server: ESF
+Vary: Origin
+Vary: X-Origin
+Vary: Referer
+X-Content-Type-Options: nosniff
+X-Frame-Options: SAMEORIGIN
+X-Xss-Protection: 0
+
+{
```
Comment (maqiuyujoyce) [inline pkg/test/resourcefixture/testdata/basic/networksecurity/v1alpha1/networksecurityfirewallendpointassociation/firewallendpointassociation-basic/_http_mock.log:536]: It's probably an edge case and out of scope, though the real GCP response for firewall endpoint seems to be quite different:

```
{
  "associatedNetworks": [
    "projects/${projectNumber}/global/networks/${networkID}"
  ],
  "associations": [
    {
      "name": "projects/${projectNumber}/locations/us-central1-a/firewallEndpointAssociations/fwea-${uniqueId}",
      "network": "projects/${projectNumber}/global/networks/${networkID}"
    }
  ],
  "billingProjectId": "${projectId}",
  "createTime": "2024-04-01T12:34:56.123456Z",
  "endpointSettings": {
    "jumboFramesEnabled": true
  },
  "labels": {
    "foo": "bar"
  },
  "name": "projects/${projectId}/locations/us-central1-a/firewallEndpoints/fwe-${uniqueId}",
  "reconciling": true,
  "state": 2,
  "updateTime": "2024-04-01T12:34:56.123456Z"
}
```

**Review by maqiuyujoyce — COMMENTED** [review]
Mostly just nits, except that could you help me understand the change in dev/tasks/build-release-bundle?

File: dev/tasks/build-release-bundle (hunk 6 above)
Comment (gemmahou) [reply dev/tasks/build-release-bundle:202]: NetworkSecurityFirewallEndpoint needs to be removed from the blocklist, it's verified in a previous PR: #12350

File: pkg/test/resourcefixture/testdata/basic/networksecurity/v1alpha1/networksecurityfirewallendpointassociation/firewallendpointassociation-basic/_http_mock.log (hunk 7 above)
Comment (gemmahou) [reply pkg/test/resourcefixture/testdata/basic/networksecurity/v1alpha1/networksecurityfirewallendpointassociation/firewallendpointassociation-basic/_http_mock.log:223]: Updated mock to not populate Labels field in operation

File: pkg/test/resourcefixture/testdata/basic/networksecurity/v1alpha1/networksecurityfirewallendpointassociation/firewallendpointassociation-basic/_http_mock.log (hunk 8 above)
Comment (gemmahou) [reply pkg/test/resourcefixture/testdata/basic/networksecurity/v1alpha1/networksecurityfirewallendpointassociation/firewallendpointassociation-basic/_http_mock.log:514]: Updated mock.

File: pkg/test/resourcefixture/testdata/basic/networksecurity/v1alpha1/networksecurityfirewallendpointassociation/firewallendpointassociation-basic/_http_mock.log (hunk 9 above)
Comment (gemmahou) [reply pkg/test/resourcefixture/testdata/basic/networksecurity/v1alpha1/networksecurityfirewallendpointassociation/firewallendpointassociation-basic/_http_mock.log:536]: Yes, the two additional fields associatedNetworks and associations are output-only, when a FirewallEndpointAssociation added to the FirewallEndpoint.

I updated mock to match the real.

**Review by maqiuyujoyce — APPROVED** [review]
/lgtm
